packages/flutter/lib/src/widgets/widget_state.dart

```diff
@@ -0,0 +1,653 @@
+/// state color property represents a color which depends on
+/// a widget's "interactive state". This state is represented as a
+/// [Set] of [WidgetState]s, like [WidgetState.pressed],
+/// [WidgetState.focused] and [WidgetState.hovered].
+///
+/// [WidgetStateColor] should only be used with widgets that document
+/// their support, like [TimePickerThemeData.dayPeriodColor].
+///
+/// To use a [WidgetStateColor], you can either:
+///   1. Create a subclass of [WidgetStateColor] and implement the abstract `resolve` method.
+///   2. Use [WidgetStateColor.resolveWith] and pass in a callback that
+///      will be used to resolve the color in the given states.
+///
+/// If a [WidgetStateColor] is used for a property or a parameter that doesn't
+/// support resolving [WidgetStateProperty<Color>]s, then its default color
+/// value will be used for all states.
+///
+/// To define a `const` [WidgetStateColor], you'll need to extend
+/// [WidgetStateColor] and override its [resolve] method. You'll also need
+/// to provide a `defaultValue` to the super constructor, so that we can know
+/// at compile-time what its default color is.
+///
+/// {@tool snippet}
+///
+/// This example defines a [WidgetStateColor] with a const constructor.
+///
+/// ```dart
+/// class MyColor extends WidgetStateColor {
+///   const MyColor() : super(_defaultColor);
+///
+///   static const int _defaultColor = 0xcafefeed;
+///   static const int _pressedColor = 0xdeadbeef;
+///
+///   @override
+///   Color resolve(Set<WidgetState> states) {
+///     if (states.contains(WidgetState.pressed)) {
+///       return const Color(_pressedColor);
+///     }
+///     return const Color(_defaultColor);
+///   }
+/// }
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [MaterialStateColor], the Material specific version of `WidgetStateColor`.
+abstract class WidgetStateColor extends Color implements WidgetStateProperty<Color> {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const WidgetStateColor(super.defaultValue);
+
+  /// Creates a [WidgetStateColor] from a [WidgetPropertyResolver<Color>]
+  /// callback function.
+  ///
+  /// If used as a regular color, the color resolved in the default state (the
+  /// empty set of states) will be used.
+  ///
+  /// The given callback parameter must return a non-null color in the default
+  /// state.
+  static WidgetStateColor resolveWith(WidgetPropertyResolver<Color> callback) => _WidgetStateColor(callback);
+
+  /// Returns a [Color] that's to be used when a component is in the specified
+  /// state.
+  @override
+  Color resolve(Set<WidgetState> states);
+
+  /// A constant whose value is transparent for all states.
+  static const WidgetStateColor transparent = _WidgetStateColorTransparent();
+}
+
+/// A [WidgetStateColor] created from a [WidgetPropertyResolver<Color>]
```

Comment:
It would be OK to leave out the doc for this private class, since its pretty simple and obvious.